bus-mapping/src/evm/opcodes/call.rs

```diff
@@ -0,0 +1,250 @@
+use super::Opcode;
+use crate::circuit_input_builder::CircuitInputStateRef;
+use crate::eth_types::{GethExecStep, ToWord};
+use crate::operation::{
+    AccountField, AccountOp, CallContextField, CallContextOp, StackOp,
+    TxAccessListAccountOp,
+};
+use crate::{operation::RW, Error};
+
+/// Placeholder structure used to implement [`Opcode`] trait over it
+/// corresponding to the `OpcodeId::DUP*` `OpcodeId`.
+#[derive(Debug, Copy, Clone)]
+pub(crate) struct Call;
+
+impl Opcode for Call {
+    fn gen_associated_ops(
+        state: &mut CircuitInputStateRef,
+        steps: &[GethExecStep],
+    ) -> Result<(), Error> {
+        let tx_id = 1;
+        let caller_call = state.caller_call().clone();
+        let call = state.call().clone();
+        let step = &steps[0];
+        let next_step = &steps[1];
+
+        for (field, value) in [
+            (CallContextField::TxId, tx_id.into()),
+            (CallContextField::RwCounterEndOfReversion, 0.into()),
+            (
+                CallContextField::IsPersistent,
+                (caller_call.is_persistent as u64).into(),
+            ),
+            (
+                CallContextField::CallerAddress,
+                caller_call.address.to_word(),
+            ),
+            (
+                CallContextField::IsStatic,
+                (caller_call.is_static as u64).into(),
+            ),
+            (CallContextField::Depth, caller_call.depth.into()),
+        ] {
+            state.push_op(CallContextOp {
+                rw: RW::READ,
+                call_id: caller_call.call_id,
+                field,
+                value,
+            });
+        }
+
+        for i in 0..7 {
+            state.push_op(StackOp {
+                rw: RW::READ,
+                call_id: caller_call.call_id,
+                address: step.stack.nth_last_filled(i),
+                value: step.stack.nth_last(i)?,
+            });
+        }
+        state.push_op(StackOp {
+            rw: RW::WRITE,
+            call_id: caller_call.call_id,
+            address: step.stack.nth_last_filled(6),
+            value: (call.is_success as u64).into(),
+        });
+
+        let value_prev = state.sdb.add_account_to_access_list(call.address);
+        state.push_op(TxAccessListAccountOp {
+            tx_id,
+            address: call.address,
+            value: true,
+            value_prev,
+        });
+        state
+            .caller_ctx_mut()
+            .push_op_reverted(TxAccessListAccountOp {
+                tx_id,
+                address: call.address,
+                value: value_prev,
+                value_prev: true,
+            });
+
+        for (field, value) in [
```

Comment:
> So before this point we where generating associated ops in the context where CALL is executed, and after this point it's going to be the context which CALL creates right?

Yes exactly, but all these ops happen in the same step.

> I think we should mask this logic behind two different functions so that it's much easier to see the context switch from CALL to the others.

For sure, I also did it in the spec like [this](https://github.com/appliedzkp/zkevm-specs/pull/82/files#diff-6e26ef25d75393b1847a36d212141d7877188735f4c74cb3f371a92466ef6538R138), maybe all these complicated things should be refactored to be methods of `CircuitInputBuidlerRef`.

Reply:
> For sure, I also did it in the spec like [this](https://github.com/appliedzkp/zkevm-specs/pull/82/files#diff-6e26ef25d75393b1847a36d212141d7877188735f4c74cb3f371a92466ef6538R138), maybe all these complicated things should be refactored to be methods of `CircuitInputBuidlerRef`.
I think adding methods to `CircuitInputBuidlerRef` for some of the stuff going on here is worth exploring.  We can move this to an issue.